protocol/x/clob/keeper/stateful_order_state.go

```diff
@@ -175,23 +175,67 @@ func (k Keeper) DeleteLongTermOrderPlacement(
 	)
 }
 
-// GetStatefulOrdersTimeSlice gets a slice of stateful order IDs that expire at `goodTilBlockTime`,
-// sorted by order ID.
-func (k Keeper) GetStatefulOrdersTimeSlice(ctx sdk.Context, goodTilBlockTime time.Time) (
+// GetStatefulOrderIdExpirations gets a slice stateful order IDs that expire at `goodTilBlockTime`,
+// sorted by order ID state key.
+func (k Keeper) GetStatefulOrderIdExpirations(ctx sdk.Context, goodTilBlockTime time.Time) (
 	orderIds []types.OrderId,
 ) {
-	store := k.getStatefulOrdersTimeSliceStore(ctx)
-	statefulOrdersTimeSliceBytes := store.Get(sdk.FormatTimeBytes(goodTilBlockTime))
+	return k.GetOrderIds(
+		ctx,
+		ctx.KVStore(k.storeKey),
+		fmt.Sprintf(types.StatefulOrdersExpirationsKeyPrefix, sdk.FormatTimeString(goodTilBlockTime)),
+	)
+}
 
-	// If there are no stateful orders that expire at this block, then return an empty slice.
-	if statefulOrdersTimeSliceBytes == nil {
-		return []types.OrderId{}
-	}
+// RemoveStatefulOrderIdExpiration removes a stateful order id expiration for an order id and
+// `goodTilBlockTime`.
+func (k Keeper) RemoveStatefulOrderIdExpiration(
+	ctx sdk.Context,
+	goodTilBlockTime time.Time,
+	orderId types.OrderId,
+) {
+	k.RemoveUnorderedOrderId(
+		ctx,
+		ctx.KVStore(k.storeKey),
+		fmt.Sprintf(types.StatefulOrdersExpirationsKeyPrefix, sdk.FormatTimeString(goodTilBlockTime)),
+		orderId,
+	)
+}
 
-	var longTermOrders types.StatefulOrderTimeSliceValue
-	k.cdc.MustUnmarshal(statefulOrdersTimeSliceBytes, &longTermOrders)
+// AddStatefulOrderIdExpiration adds a stateful order id expiration for an order id and
+// `goodTilBlockTime`.
+func (k Keeper) AddStatefulOrderIdExpiration(
+	ctx sdk.Context,
+	goodTilBlockTime time.Time,
+	orderId types.OrderId,
+) {
+	k.SetUnorderedOrderId(
+		ctx,
+		ctx.KVStore(k.storeKey),
+		fmt.Sprintf(types.StatefulOrdersExpirationsKeyPrefix, sdk.FormatTimeString(goodTilBlockTime)),
+		orderId,
+	)
+}
 
-	return longTermOrders.OrderIds
+// RemoveExpiredStatefulOrders removes the stateful order id expirations up to `blockTime` and
+// returns the removed order ids as a slice.
+func (k Keeper) RemoveExpiredStatefulOrders(ctx sdk.Context, blockTime time.Time) (
+	expiredOrderIds []types.OrderId,
+) {
+	expiredOrderIds = make([]types.OrderId, 0)
+	store := ctx.KVStore(k.storeKey)
+	start := []byte(fmt.Sprintf(types.StatefulOrdersExpirationsKeyPrefix, sdk.FormatTimeString(time.Time{})))
+	end := []byte(fmt.Sprintf(types.StatefulOrdersExpirationsKeyPrefix, sdk.FormatTimeString(blockTime)))
+	// Increment the byte representation of `/` at the end of the prefix so that blockTime is included in the iteration
+	end[len(end)-1] += 1
```

Comment:
can this overflow? not sure what formatted time string this is using

there is a util function in sdk store types, which might be better
```
// InclusiveEndBytes returns the []byte that would end a
// range query such that the input would be included
func InclusiveEndBytes(inclusiveBytes []byte) []byte {
	return append(inclusiveBytes, byte(0x00))
}
```

Reply:
Since the key prefix is `Exp/%s:`, if you do that, you'd get `Exp/{time}:\00` as the exclusive end, which would not cover all the keys that the orders are stored it. For example `"Exp/{time}:{order_id}" would not get iterated over.

The idea is to increment the byte representation of `:` by 1 so that all order ids at the ending time would be included.

If you can think of better ways of doing this, I'm open to it.

Reply:
Thinking out loud - could we define another `StatefulOrdersExpirationsKeyPrefixExclusiveEnd = "Exp/%s;"`, and do `end := []byte(fmt.Sprintf(types.StatefulOrdersExpirationsKeyPrefixExclusiveEnd, sdk.FormatTimeString(blockTime)))`

Reply:
Yeah that's pretty much the same thing though.

Reply:
there is also another function that we can use called `PrefixEndBytes`

```
// PrefixEndBytes returns the []byte that would end a
// range query for all []byte with a certain prefix
// Deals with last byte of prefix being FF without overflowing
func PrefixEndBytes(prefix []byte) []byte {
	if len(prefix) == 0 {
		return nil
	}

	end := make([]byte, len(prefix))
	copy(end, prefix)

	for {
		if end[len(end)-1] != byte(255) {
			end[len(end)-1]++
			break
		}

		end = end[:len(end)-1]

		if len(end) == 0 {
			end = nil
			break
		}
	}

	return end
}
```

Reply:
which does what you are trying to do, and handles overflow